src/utils/changelog.ts

```diff
@@ -0,0 +1,396 @@
+    );
+  }
+
+  return changesetLoc ? extractChangeset(markdown, changesetLoc) : null;
+}
+
+/**
+ * Removes a given changeset from the provided markdown and returns the modified markdown
+ * @param markdown The markdown containing the changeset
+ * @param header The header of the changeset to-be-removed
+ * @returns The markdown string without the changeset with the provided header
+ */
+export function removeChangeset(markdown: string, header: string): string {
+  const location = locateChangeset(markdown, match => match === header);
+  if (!location) {
+    return markdown;
+  }
+
+  const start = location.start.index;
+  const end = location.end?.index ?? markdown.length;
+  return markdown.slice(0, start) + markdown.slice(end);
+}
+
+/**
+ * Prepends a changeset to the provided markdown text and returns the result.
+ * It tries to prepend before the first ever changeset header, to keep any
+ * higher-level content intact and in order. If none found, then the changeset
+ * is *appended* instead.
+ *
+ * @param markdown The markdown that will be prepended
+ * @param changeset The changeset data to prepend to
+ * @returns The markdown string with the changeset prepedend before the top-most
+ *          existing changeset.
+ */
+export function prependChangeset(
+  markdown: string,
+  changeset: Changeset
+): string {
+  // Try to locate the top-most non-empty header, no matter what is inside
+  const { start, padding } = locateChangeset(markdown, Boolean) || {
+    padding: '',
+  };
+  const body = changeset.body || `${padding}${DEFAULT_CHANGESET_BODY}`;
+  let header;
+  if (start?.[3]) {
+    const underline = new Array(changeset.name.length + 1).join('-');
+    header = `${changeset.name}\n${underline}`;
+  } else {
+    header = `## ${changeset.name}`;
+  }
+  const newSection = `${padding}${header}\n\n${body.replace(
+    /^/gm,
+    padding
+  )}\n\n`;
+  const startIdx = start?.index ?? markdown.length;
+
+  return markdown.slice(0, startIdx) + newSection + markdown.slice(startIdx);
+}
+
+interface Commit {
+  hash: string;
+  title: string;
+  hasPRinTitle: boolean;
+  pr: string | null;
+  milestone: string | null;
+}
+
+interface Milestone {
+  title?: string;
+  description?: string;
+  state?: 'OPEN' | 'CLOSED';
+}
+type MilestoneWithPRs = Milestone & {
+  prs: string[];
+};
+
+// This is set to 8 since GitHub and GitLab prefer that over the default 7 to
+// avoid collisions.
+const SHORT_SHA_LENGTH = 8;
+function formatCommit(commit: Commit): string {
+  let text = `- ${commit.title}`;
+  if (!commit.hasPRinTitle) {
+    const link = commit.pr
+      ? `#${commit.pr}`
+      : commit.hash.slice(0, SHORT_SHA_LENGTH);
+    text = `${text} (${link})`;
+  }
+  return text;
+}
+
+export async function generateChangesetFromGit(
+  git: SimpleGit,
+  rev: string
+): Promise<string> {
+  const gitCommits = await getChangesSince(git, rev);
+
+  const githubCommits = await getMilestoneAndPRFromCommits(
```

Comment:
nit: `getMilestonesAndPRsFromCommits` since we get plural milestones and PRs (right?) and we have plural in `getMilestonesDetails`

Reply:
Or since milestones are nested under PRs: `getPRsWithMilestonesFromCommits`?

Reply:
The naming is due to the `commit -> PR -> milestone` mapping. 1 commit to 1 PR to 1 Milestone. I guess making the other one singular is the right way to go here.